### PR TITLE
Transaction code porting from Framwork for TransactionScope / TransactionBinding support

### PR DESCRIPTION
**Pull request by geleems**
Code related to SQL transaction was ported from NET Framework for supporting TransactionScope and TransactionBinding.

### CODE REVIEW

File: src/Common/src/System/Data/Common/AdapterUtil.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/Common/src/System/Data/Common/AdapterUtil.cs]: Why do you need this alias here? Are there duplicate Transaction classes in this scope?

File: src/Common/src/System/Data/Common/AdapterUtil.cs [hunk 2]
```diff
@@ -428,5 +434,13 @@ internal static Exception InvalidSeekOrigin(string parameterName)
         {
             return ArgumentOutOfRange(SR.ADP_InvalidSeekOrigin, parameterName);
         }
+
+        internal static readonly bool IsWindowsNT = (PlatformID.Win32NT == Environment.OSVersion.Platform);
+        internal static readonly bool IsPlatformNT5 = (ADP.IsWindowsNT && (Environment.OSVersion.Version.Major >= 5));
+
+        static internal void SetCurrentTransaction(SysTx.Transaction transaction)
```
Comment (corivera) [inline src/Common/src/System/Data/Common/AdapterUtil.cs]: internal static

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 3]
```diff
@@ -113,18 +113,23 @@
     <Compile Include="System\Data\SqlClient\SqlCommandSet.cs" />
     <Compile Include="System\Data\SqlClient\SqlConnection.cs" />
     <Compile Include="System\Data\SqlClient\SqlConnectionFactory.cs" />
-    <Compile Include="System\Data\SqlClient\SqlConnectionHelper.cs" />
+    <Compile Include="System\Data\SqlClient\SqlConnectionHelper.cs">
+      <SubType>Component</SubType>
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: Remove these component tags from the csproj

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs [hunk 4]
```diff
@@ -28,7 +29,7 @@ override public string ServerVersion
             }
         }
 
-        override protected void Activate()
+        override protected void Activate(SysTx.Transaction transaction)
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs]: protected override

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs [hunk 5]
```diff
@@ -53,6 +54,10 @@ override protected void Deactivate()
             throw ADP.ClosedConnectionError();
         }
 
+        override public void EnlistTransaction(SysTx.Transaction transaction)
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs]: public override

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs [hunk 6]
```diff
@@ -1015,7 +1020,7 @@ private bool TryOpen(TaskCompletionSource<DbConnectionInternal> retry)
             // Fail Fast in case an application is trying to enlist the SqlConnection in a Transaction Scope.
             if (connectionOptions.Enlist && ADP.GetCurrentTransaction() != null)
             {
-                throw ADP.AmbientTransactionIsNotSupported();
+                //throw ADP.AmbientTransactionIsNotSupported();
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: Do we still need this if case? Is it present in Framework?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionHelper.cs [hunk 7]
```diff
@@ -154,6 +155,41 @@ override protected void Dispose(bool disposing)
 
         partial void RepairInnerConnection();
 
+        override public void EnlistTransaction(SysTx.Transaction transaction)
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionHelper.cs]: public override

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs [hunk 8]
```diff
@@ -153,6 +153,17 @@ internal static class TYPESYSTEMVERSION
             internal const string SQL_Server_2012 = "SQL Server 2012";
         }
 
+        internal enum TransactionBindingEnum
+        {
+            ImplicitUnbind,
+            ExplicitUnbind
+        }
+
+        internal static class TRANSACIONBINDING
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs]: typo in transaction

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs (hunk 6 above)
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: This should be removed. I had added this for 2.0

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs [hunk 9]
```diff
@@ -342,6 +356,23 @@ internal SqlConnectionString(string connectionString) : base(connectionString, G
                 throw ADP.InvalidConnectionOptionValue(KEY.Type_System_Version);
             }
 
+            if (ADP.IsEmpty(transactionBindingString))
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs]: string.IsNullOrEmpty() ?

File: src/Common/src/System/Data/Common/AdapterUtil.cs [hunk 10]
```diff
@@ -414,6 +415,11 @@ internal static Exception InvalidDataLength(long length)
 
         internal static bool IsEmptyArray(string[] array) => (null == array) || (0 == array.Length);
 
+        internal static bool IsEmpty(string str)
```
Comment (saurabh500) [inline src/Common/src/System/Data/Common/AdapterUtil.cs]: This method should be removed. Use sting.IsNullOrEmpty

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs [hunk 11]
```diff
@@ -510,7 +544,7 @@ internal SqlConnectionString(SqlConnectionString connectionOptions, string dataS
                     { SYNONYM.User, KEY.User_ID },
                     { SYNONYM.WSID, KEY.Workstation_Id }
                 };
-                Debug.Assert(synonyms.Count == count, "incorrect initial ParseSynonyms size");
+                //Debug.Assert(synonyms.Count == count, "incorrect initial ParseSynonyms size");
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionString.cs]: Is this Debug Assert still needed?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs [hunk 12]
```diff
@@ -157,6 +159,10 @@ internal static class SQL
         //
         // SQL.Connection
         //
+        static internal Exception CannotGetDTCAddress()
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs]: internal static

Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs]: Same for other methods in this class

**Comment by corivera** [comment]
Looks like the tests are failing because they're expecting SqlExceptions and NotSupportedExceptions from Transactions not being implemented.

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs [hunk 13]
```diff
@@ -9,6 +9,7 @@
 using System.Data.Common;
 using System.Diagnostics;
 using System.Threading.Tasks;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionClosed.cs]: Is this alias still needed?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 14]
```diff
@@ -334,13 +336,25 @@ public SqlNotificationRequest Notification
             }
         }
 
+        public bool NotificationAutoEnlist
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Auto enlist is not supported in Core

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 15]
```diff
@@ -1073,7 +1087,8 @@ private int EndExecuteNonQueryInternal(IAsyncResult asyncResult)
                             bool dataReady;
                             Debug.Assert(_stateObj._syncOverAsync, "Should not attempt pends in a synchronous call");
                             bool result = _stateObj.Parser.TryRun(RunBehavior.UntilDone, this, null, null, _stateObj, out dataReady);
-                            if (!result) { throw SQL.SynchronousCallMayNotPend(); }
+                            if (!result)
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Weird indentation here

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 16]
```diff
@@ -2352,7 +2367,8 @@ private Task RunExecuteNonQueryTds(string methodName, bool async, int timeout, b
                     bool dataReady;
                     Debug.Assert(_stateObj._syncOverAsync, "Should not attempt pends in a synchronous call");
                     bool result = _stateObj.Parser.TryRun(RunBehavior.UntilDone, this, null, null, _stateObj, out dataReady);
-                    if (!result) { throw SQL.SynchronousCallMayNotPend(); }
+                    if (!result)
+                    { throw SQL.SynchronousCallMayNotPend(); }
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Same as above. Just leave it inlined like before

Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: (and for all the other occurrences of this below)

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionHelper.cs [hunk 17]
```diff
@@ -10,6 +10,7 @@
 using System.Data.ProviderBase;
 using System.Diagnostics;
 using System.Threading;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionHelper.cs]: Is this still needed?

Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionHelper.cs]: Same for other occurrences of this alias

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 18]
```diff
@@ -1835,8 +1848,15 @@ internal bool TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataRead
                                                 _statistics.SafeIncrement(ref _statistics._transactions);
                                             }
                                             _statisticsIsInTransaction = true;
+                                            _retainedTransactionId = SqlInternalTransaction.NullTransactionId;
                                             break;
+                                        case TdsEnums.ENV_DEFECTDTC:
+                                        case TdsEnums.ENV_TRANSACTIONENDED:
                                         case TdsEnums.ENV_COMMITTRAN:
+                                            // SQLHOT 483
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: You can remove these internal error number references

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.SqlClient.cs [hunk 19]
```diff
@@ -897,5 +905,13 @@ internal static Exception NoStoredProcedureExists(string sproc)
         {
             return InvalidOperation(SR.GetString(SR.ADP_NoStoredProcedureExists, sproc));
         }
+
+        //
+        // DbProviderException
+        //
+        static internal InvalidOperationException TransactionCompletedButNotDisposed()
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.SqlClient.cs]: internal static

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionInternal.cs [hunk 20]
```diff
@@ -344,7 +521,7 @@ internal void DeactivateConnection()
 
 #if DEBUG
             int activateCount = Interlocked.Decrement(ref _activateCount);
-            Debug.Assert(0 == activateCount, "activated multiple times?");
+            //Debug.Assert(0 == activateCount, "activated multiple times?");
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionInternal.cs]: Is this debug assert still needed?

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionInternal.cs [hunk 21]
```diff
@@ -11,6 +11,7 @@
 using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionInternal.cs]: Unnecessary using alias

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionPool.cs [hunk 22]
```diff
@@ -12,6 +12,7 @@
 using System.Threading;
 using System.Threading.Tasks;
 using System.Collections.Concurrent;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionPool.cs]: Unnecessary using alias

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDelegatedTransaction.cs [hunk 23]
```diff
@@ -0,0 +1,466 @@
+﻿using System.Data.Common;
+using System.Data.SqlClient;
+using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDelegatedTransaction.cs]: Unnecessary alias

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnection.cs [hunk 24]
```diff
@@ -9,13 +9,21 @@
 using System.Data.Common;
 using System.Data.ProviderBase;
 using System.Diagnostics;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnection.cs]: Unnecessary alias

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs [hunk 25]
```diff
@@ -14,6 +14,7 @@
 using System.Threading;
 using System.Diagnostics.CodeAnalysis;
 using System.Threading.Tasks;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs]: Unnecessary alias

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs [hunk 26]
```diff
@@ -10,6 +10,8 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Reflection;
+using SysTx = System.Transactions;
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs]: Unnecessary alias

**Comment by geleems** [comment]
@dotnet-bot test Windows x64 Debug Build please.

**Pull request merged**